runtime/sam/op/meta/sequence.go

```diff
@@ -92,6 +94,62 @@ func (s *SequenceScanner) close(err error) {
 	s.done = true
 }
 
+type SearchScanner struct {
+	filter   zbuf.Filter
+	parent   Searcher
+	pool     *lake.Pool
+	progress *zbuf.Progress
+	rctx     *runtime.Context
+	scanner  zbuf.Puller
+}
+
+type Searcher interface {
+	Pull(bool) (*data.Object, *vector.Bool, error)
+}
+
+func NewSearchScanner(rctx *runtime.Context, parent Searcher, pool *lake.Pool, filter zbuf.Filter, progress *zbuf.Progress) *SearchScanner {
+	return &SearchScanner{
+		filter:   filter,
+		parent:   parent,
+		pool:     pool,
+		progress: progress,
+		rctx:     rctx,
+	}
+}
+
+func (s *SearchScanner) Pull(done bool) (zbuf.Batch, error) {
+	if done {
+		var err error
+		if s.scanner != nil {
+			_, err = s.scanner.Pull(true)
+			s.scanner = nil
+		}
+		return nil, err
+	}
+	for {
+		if s.scanner == nil {
+			o, b, err := s.parent.Pull(done)
+			if b == nil || err != nil {
+				return nil, err
+			}
+			ranges, err := data.RangeFromBitVector(s.rctx.Context, s.pool.Storage(), s.pool.DataPath, o, b)
+			if err != nil {
+				return nil, err
+			}
+			s.scanner, err = newObjectScanner(s.rctx.Context, s.rctx.Zctx, s.pool, o, ranges, s.filter, s.progress)
+			if err != nil {
+				return nil, err
+			}
+		}
+		batch, err := s.scanner.Pull(false)
+		if batch == nil {
+			s.scanner = nil
+			return nil, err
+		}
+		return batch, nil
```

Comment:
This won't always return a non-nil error.
```suggestion
		if batch == nil {
			s.scanner = nil
		}
		return batch, err
```